inc/saitypes.h

```diff
@@ -941,6 +941,161 @@ typedef struct _sai_port_eye_values_list_t
+     */
+    SAI_NATIVE_HASH_FIELD_DST_IPV4_ADDR,
+
+    /**
+     * @brief Native hash field source IPv6 address
+     *
+     * Also, refers to the outer source IP
+     * in case for encapsulated packets
+     */
+    SAI_NATIVE_HASH_FIELD_SRC_IPV6_ADDR,
+
+    /**
+     * @brief Native hash field destination IPv6 address
+     *
+     * Also, refers to the outer source IP
+     * in case for encapsulated packets
+     */
+    SAI_NATIVE_HASH_FIELD_DST_IPV6_ADDR,
+
+    /** Native hash field inner source IPv4 address */
+    SAI_NATIVE_HASH_FIELD_INNER_SRC_IPV4_ADDR,
+
+    /** Native hash field inner destination IPv4 address */
+    SAI_NATIVE_HASH_FIELD_INNER_DST_IPV4_ADDR,
+
+    /** Native hash field inner source IPv6 address */
+    SAI_NATIVE_HASH_FIELD_INNER_SRC_IPV6_ADDR,
+
+    /** Native hash field inner destination IPv6 address */
+    SAI_NATIVE_HASH_FIELD_INNER_DST_IPV6_ADDR,
+
+    /** Native hash field vlan id */
+    SAI_NATIVE_HASH_FIELD_VLAN_ID,
+
+    /** Native hash field IP protocol */
+    SAI_NATIVE_HASH_FIELD_IP_PROTOCOL,
+
+    /** Native hash field Ethernet type */
+    SAI_NATIVE_HASH_FIELD_ETHERTYPE,
+
+    /** Native hash field L4 source port */
+    SAI_NATIVE_HASH_FIELD_L4_SRC_PORT,
+
+    /** Native hash field L4 destination port */
+    SAI_NATIVE_HASH_FIELD_L4_DST_PORT,
+
+    /** Native hash field source MAC */
+    SAI_NATIVE_HASH_FIELD_SRC_MAC,
+
+    /** Native hash field destination MAC */
+    SAI_NATIVE_HASH_FIELD_DST_MAC,
+
+    /** Native hash field source port */
+    SAI_NATIVE_HASH_FIELD_IN_PORT,
+
+    /** Native hash field inner IP protocol */
+    SAI_NATIVE_HASH_FIELD_INNER_IP_PROTOCOL,
+
+    /** Native hash field inner Ethernet type */
+    SAI_NATIVE_HASH_FIELD_INNER_ETHERTYPE,
+
+    /** Native hash field inner L4 source port */
+    SAI_NATIVE_HASH_FIELD_INNER_L4_SRC_PORT,
+
+    /** Native hash field inner L4 destination port */
+    SAI_NATIVE_HASH_FIELD_INNER_L4_DST_PORT,
+
+    /** Native hash field inner source MAC */
+    SAI_NATIVE_HASH_FIELD_INNER_SRC_MAC,
+
+    /** Native hash field inner destination MAC */
+    SAI_NATIVE_HASH_FIELD_INNER_DST_MAC,
+
+} sai_native_hash_field_t;
+
+/**
+ * @brief Hash field mask
+ */
+typedef union _sai_native_hash_field_mask_t
+{
+    /** Mask for byte fields */
+    sai_uint8_t u8;
+
+    /** Mask for word fields */
+    sai_uint16_t u16;
+
+    /** Mask for word fields */
+    sai_uint32_t u32;
+
+    /** Mask for MAC address */
+    sai_mac_t mac;
+
+    /** Mask for IPv4 address */
+    sai_ip4_t ip4;
+
+    /** Mask for IPv6 address */
+    sai_ip6_t ip6;
+
+} sai_native_hash_field_mask_t;
+
+/**
+ * @brief Masked hash field
+ */
+typedef struct _sai_fine_grain_hash_field
+{
+    /** Flag indicating if field is native or UDF */
+    bool is_native;
+
+    /** Native field ID */
+    sai_native_hash_field_t native_field_id;
+
+    /** Mask */
+    sai_native_hash_field_mask_t mask;
+
+    /** UDF field ID */
```

Comment:
out metadata dont support object id's in complicated structures like this, are you able to remove this field ? and add workaround in other way?

Reply:
also why you need some complicate structures ? can this entire thing be convereted to new object type ?